implementing optional interfaces for async operations / updates

chrisfarms opened the pull request:
## What

* Refactor to always use the upstream `domain.*` return types to stay closer to the upstream library
* Support wrapping the methods required for async binding (`LastBindingOperation` and `GetBinding`)
* Refactor interfaces so that you can optionally implement the various interfaces in your provider to avoid needing to create a bunch of "not-implemented" methods when you don't use them
* Allow configuring the locket retry interval so that running tests does not hang for 30s+

## Why

We need async binding in the sqs broker, and also found the return types inconsistent / messy to work with compared to the upstream types.

## Note

*this is a breaking change*

The change to the interfaces means that those implementing Providers will need to update their function signatures:

Old...

```
func (p *MyProvider) Provision(context.Context, ProvisionData) (string, string, bool, error) {
    return "", "", true, nil
}
```

New...

```
func (p *MyProvider) Provision(context.Context, ProvisionData) (*domain.ProvisionServiceSpec, error) {
    return &domain.ProvisionServiceSpec{
        IsAsync: true,
    }, nil
}
```

This change keeps the broker-base wrapper closer to the upstream usage and brings consistency to each of the method calls.

Commit:
refactoring interface to support async bind

Krenair committed:
Fix a couple of seeming mistakes in the tests, which now pass

Commit:
allow configuring locket retry interval

test runs are very slow because there is a 1s sleep during attempting
locket retry logic

allow configuring the retry interval so we can speed up the tests

the default value (when no retry interval is explicitly set) remains the
same at 1s so this should have no impact on anything other than the
tests

Krenair committed:
more tests, make counterfeiter work

Krenair committed:
Sort out a bunch of coverage issues, test more things

Krenair committed:
Fix LastBindingOperation parameter type

Krenair committed:
Implement LastBindingOperation and GetBinding in the broker tester

philandstuff commented on broker/broker.go:
```diff
@@ -16,18 +16,25 @@ import (
 	logger       lager.Logger
 	LocketClient locket_models.LocketClient
 }
 
```
```suggestion

var _ domain.ServiceBroker = &Broker{}

```

It wasn't clear to me on first reading why this code had the parameters it did (for example, why does `Services()` take a context, and then not use it?).  This line would a) document that intention, and b) get the compiler to enforce that a *Broker is a domain.ServiceBroker.

Commit:
Add type check that Broker implements domain.ServiceBroker

Co-authored-by: Philip Potter <philip.potter@digital.cabinet-office.gov.uk>

Commit:
generate fakes

philandstuff commented on broker/broker.go:
```diff
@@ -151,22 +177,29 @@ func (b *Broker) Deprovision(
 		return domain.DeprovisionServiceSpec{}, err
 	}
+	if res == nil {
+		return domain.DeprovisionServiceSpec{}, ErrNilResponse
```
if `nil` is not a valid response, can we change the `Provisioner` interface to not return a pointer? then we can ditch this and similar conditions.  It also looks more similar to the upstream ServiceBroker interface so people don't get confused when here it's pointers and elsewhere it's raw structs.

chrisfarms replied:
the intention here is to make the return types nicer to work with. So here it is only be an error if the response is `nil` AND `err == nil` ....making it return a concrete type means all the error cases end up being:

```
return domain.ThingIDontCareAbout{}, errors.New("my error")
```

this isn't really wrong, but it can lead to some issues and in our case this actually caught a bug. we had some code like this...

```
if err != nil {
   return domain.Thing{}, nil
}
```

It should have been disguarding the Thing and returning the error... but because that is a totally valid return type, it went unoticed.

with the pointer return type...

```
if err != nil {
    return nil, nil
}
```

... would have stood out as wrong and would also have caused an error at runtime

philandstuff replied:
makes sense. Marking resolved.

philandstuff commented on broker/broker.go:
```diff
@@ -241,26 +284,66 @@ func (b *Broker) Unbind(
+		"binding-id":  bindingID,
+	})
+
+	providerCtx, cancelFunc := context.WithTimeout(ctx, 60*time.Second)
```
is it worth extracting `60*time.Second` into a constant?

philandstuff commented on broker/broker_test.go:
```diff
@@ -890,6 +1137,118 @@ var _ = Describe("Broker", func() {
+
+			_, hasDeadline := receivedContext.Deadline()
+
+			Expect(hasDeadline).To(BeTrue())
```
ooh did not know you could test deadlines like this. :nice:

philandstuff commented on broker/locket.go:
```diff
@@ -3,19 +3,27 @@ package broker
 
+var (
+	// DefaultLocketRetryInterval is the time between retries while waiting
+	// for successful locket response
```
hooray for godoc \o/

chrisfarms replied:
one down ... several hundred to go :champagne:

philandstuff commented on provider/interface.go:
```diff
@@ -6,12 +6,38 @@ import (
 
-//go:generate counterfeiter -o fakes/fake_service_provider.go . ServiceProvider
-type ServiceProvider interface {
-	Provision(context.Context, ProvisionData) (dashboardURL, operationData string, isAsync bool, err error)
```
really happy to see these bag-of-strings return types go 🚮  👍

chrisfarms replied:
me too ... 

```
return "", "", "", true, blah // yuk
```

philandstuff commented on provider/interface.go:
```diff
@@ -6,12 +6,38 @@ import (
-	Unbind(context.Context, UnbindData) (unbinding domain.UnbindSpec, err error)
-	Update(context.Context, UpdateData) (operationData string, isAsync bool, err error)
-	LastOperation(context.Context, LastOperationData) (state domain.LastOperationState, description string, err error)
+type Provider interface{}
```
are literally all operations optional? (I mean, I guess the answer is "yes" in the case of the observe service broker, but it's not exactly a great example.)

I would have thought that Provision/Deprovision would be minimum required operations?

chrisfarms replied:
Yeah I'm on the fense about that one ... I think `Provider` should probably contain `Provisioner` ... I don't like the empty interface either

chrisfarms replied:
see other comment on combining interfaces

philandstuff commented on provider/interface.go:
```diff
@@ -6,12 +6,38 @@ import (
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -o fakes/fake_binder.go . Binder
+type Binder interface {
+	Bind(context.Context, BindData) (binding *domain.Binding, err error)
```
Does it make sense to have a Binder which is not also a Provisioner?
```suggestion
	Provisioner
	Bind(context.Context, BindData) (binding *domain.Binding, err error)
```

chrisfarms replied:
Agreed ... I was also considering these too (some kind of internal struggle between "interfaces just do one thing" and "this is the real world, get over it" :smile: )

chrisfarms replied:
ok, this gets a bit complicated and ends up a bit rabit holey ... I attempted to make Provider==Provisioner and embed the relevent types together, however it is quite a big change because:

* Avoiding overlapping interfaces becomes a problem (ie you want to implement Provider and AsyncBinder but NOT AsyncProvider)
* Some of the tests become significant;y harder to write (requiring major refactoring to accommodate - especially the api_tests which as mainly un-changed by this PR)

It's possible, but after a good 40mins I am backing out as the changes just keep piling up and I don't think the reward is worth the effort.

If you feel very strongly about these interface changes, then I can plough on, but it was alot more effort than expected.

philandstuff commented on testing/broker_tester_test.go:
```diff
@@ -8,7 +8,7 @@ import (
 )
 
-//go:generate counterfeiter -o fakes/fake_http_handler.go net/http.Handler
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -o fakes/fake_http_handler.go net/http.Handler
```
👀  this seems like a nice pattern.  Is go mod smart enough not to remove counterfeiter on `go mod tidy` if it appears here?

chrisfarms replied:
go mod is using the file in `tools/tool.go` to pick up the dep

chrisfarms replied:
which I just relised @Krenair hasn't committed :raising_hand:

Krenair replied:
done, sorry

philandstuff reviewed (commented):
Lots of comments, but only the first one is blocking, the rest are all praise or "that's interesting" or "did you consider"

Krenair committed:
Add missing files

Commit:
break out context timeout into const

avoid repeating the time.Second*30 for each method

named Default* in expectation that this will become configurable for
providers that need longer request timeouts
